core/src/kvs/tx.rs

```diff
@@ -84,6 +85,108 @@ where
 	pub values: Vec<(Key, Val)>,
 }
 
+pub trait Transactable {
+	/// Check if closed
+	fn closed(&self) -> bool;
+	/// Cancel a transaction
+	fn cancel(&mut self) -> impl Future<Output = Result<(), Error>>;
+	/// Commit a transaction
+	fn commit(&mut self) -> impl Future<Output = Result<(), Error>>;
+	/// Check if a key exists
+	fn exi<K>(&mut self, key: K) -> impl Future<Output = Result<bool, Error>>
+	where
+		K: Into<Key>;
+	/// Fetch a key from the database
+	fn get<K>(&mut self, key: K) -> impl Future<Output = Result<Option<Val>, Error>>
+	where
+		K: Into<Key>;
+	/// Insert or update a key in the database
+	fn set<K, V>(&mut self, key: K, val: V) -> impl Future<Output = Result<(), Error>>
+	where
+		K: Into<Key>,
+		V: Into<Val>;
+	/// Insert a key if it doesn't exist in the database
+	fn put<K, V>(
+		&mut self,
+		category: KeyCategory,
+		key: K,
+		val: V,
+	) -> impl Future<Output = Result<(), Error>>
+	where
+		K: Into<Key>,
+		V: Into<Val>;
+	/// Insert a key if it doesn't exist in the database
+	fn putc<K, V>(
+		&mut self,
+		key: K,
+		val: V,
+		chk: Option<V>,
+	) -> impl Future<Output = Result<(), Error>>
+	where
+		K: Into<Key>,
+		V: Into<Val>;
+	/// Delete a key
+	fn del<K>(&mut self, key: K) -> impl Future<Output = Result<(), Error>>
+	where
+		K: Into<Key>;
+	/// Delete a key
+	fn delc<K, V>(&mut self, key: K, chk: Option<V>) -> impl Future<Output = Result<(), Error>>
+	where
+		K: Into<Key>,
+		V: Into<Val>;
+
+	#[allow(unused_variables)]
+	fn delr<K>(&mut self, rng: Range<K>, limit: u32) -> impl Future<Output = Result<(), Error>>
+	where
+		K: Into<Key>,
+	{
+		async { Err(Error::Unimplemented(String::new())) }
```

Comment:
Can we at least say what's not implemented? After this error message has been propagated or it shows up in the logs, it won't be straightforward to find out where it originated from.

Reply:
Yes, but since the first item type of the tuple for `Error::Unimplemented` is `String`, I can't just use a static string directly (`&'static str`).

Reply:
You can do `Err(Error::Unimplemented("error message here".to_owned()))` or `Err(Error::Unimplemented(String::from("error message here")))` etc.

Reply:
That means there will be a heap allocation every time if the method was not implemented. Is it really acceptable, though?

Reply:
Yes it does. Ideally this variant should be `&'static str` to avoid the heap allocation while still being able to embed a custom message but `String` is already part of the API. It's possible that there is some part of the code where the string was necessary.

Having said that this is not the happy path and I don't expect to see a lot of unimplemented errors in production, if at all. It's more important that these errors are easy to debug. Are there any implementations that don't implement these methods? If not, we could also just delete these default implementations to ensure that they are implemented.